Add vm security group operations

tzumainn opened the pull request:
Adds VM security group operations

tzumainn commented:
![screenshot from 2017-08-25 13 46 17](https://user-images.githubusercontent.com/1280178/29725929-d5ea8e3e-899b-11e7-8b43-103441f8394c.png)

tzumainn commented:
@h-kataria Thanks for the guidance, let me know if I did this right!

tzumainn commented:
@miq-bot add_label UX/Review

tzumainn commented:
@miq-bot remove_label wip

tzumainn commented:
@h-kataria I believe this is ready for review now!

dclarizio commented:
@tzumainn we relaxed the codeclimate checks, can you see if you can address any of the 18 issues that are reported?

tzumainn commented:
@dclarizio thanks for pointing that out! I took a closer look at the codeclimate issues, and I'm not sure any of them can be addressed. The duplicated code is extremely minor, and feels dangerous to pull out. It looks like I already follow the suggested angular chaining - I think the lint is wrong. And the 'unless' statements look correct to me.

Let me know if there's something specific in there you think I missed!

dclarizio commented:
@martinpovolny Can you review the CC issues?  I know we relaxed CC checks, does CC need to be rerun?

serenamarie125 commented:
@tzumainn how are they accessing this operation?

tzumainn commented:
@serenamarie125 From a single VM view, under the Configuration menu - same place where associate/disassociate floating IP and similar operations,

martinpovolny commented:
@tzumainn, @dclarizio : More that about CC issues -- yes, there are 18 -- I am concerned about creating more and more new RJS code. 

This PR introduces a lot of new technical debt. Is there a reason to do this?  Are we in a hurry? Is this a Sev  1 fix?

martinpovolny commented:
To be more specific:

 * Why is the API not used?
 * There should be no new `render :update` code at all unless you are adding a small thing to a complex controller where everything is done that way.
 * `app/controllers/mixins/actions/vm_actions/remove_security_group.rb` and `app/controllers/mixins/actions/vm_actions/add_security_group.rb` look like c-n-ped.

martinpovolny commented on app/assets/javascripts/controllers/vm_cloud/vm_cloud_remove_security_group_form_controller.js:
```diff
@@ -0,0 +1,40 @@
+  vm.resetClicked = function(angularForm) {
+    vm.CloudModel = angular.copy( vm.modelCopy );
+    angularForm.$setPristine(true);
+    miqService.miqFlash("warn", "All changes have been reset");
```
no i18n?

martinpovolny commented on app/assets/javascripts/controllers/vm_cloud/vm_cloud_remove_security_group_form_controller.js:
```diff
@@ -0,0 +1,40 @@
+ManageIQ.angular.app.controller('vmCloudRemoveSecurityGroupFormController', ['$http', 'vmCloudRemoveSecurityGroupFormId', 'miqService', function($http, vmCloudRemoveSecurityGroupFormId, miqService) {
```
We are trying to convert to Angular 1.6 so that we can move forward with Angular 2 and React. This should be a component. The partials below shoud not be created, those should be part of the component or separate componets too.

Ping @himdel, @tzumainn. http://www.codelord.net/2015/10/07/angular-2-preparation-killing-controllers/

Hyperkid123 replied:
There are **no controllers in Angular 2+**, so i completely that we should try to write any new angular **controllers** as **components**, since we will have to refactor them anyway... And that applies especially for partials, because refactoring them later after they go through several modifications is pain in the ass.

You can use [this article](https://teropa.info/blog/2015/10/18/refactoring-angular-apps-to-components.html) as reference. There are also some components already in ui-classic(and more is comming).

himdel replied:
Working on the guide, but for now, pretty much all you need to do is:

* rename `init` to `$onInit` - that gets called automagically by angular, so you no longer need to call it manually.
* move the templates under `app/views/static/` - that way, you can reference it via `templateUrl: '/static/....'` from angular
* remove any controller-dependent logic from the moved template - that just means you can't use `@record.id` in your case
* make it into a component...

```js
ManageIQ.angular.app.controller("foo", **controller**);
```

should become

```js
ManageIQ.angular.app.component("foo", {
  bindings: {
    recordId: "@",
  }
  templateUrl: "/static/foo.html.haml",
  controller: **controller**,
});
```

* then you no longer need to pass the id using `value`, you can provide it to the component as a parameter - hence the `bindings` entry - so no injecting `vmCloudRemoveSecurityGroupFormId` any more ;)
* replace the original template with just something that uses the component .. so, `%my-component{:record-id => @record.id}`, and initializes angular (so that `miq_bootstrap` stays here, not in `static` - we usually use the component name as the first arg in those cases (so you don't need an extra `#div`))
* don't leave the code in `app/assets/javascripts/controllers`, use the `components` dir for that - or if you're up to speed with the recent webpacker changes ([talk](http://talk.manageiq.org/t/webpack-er-for-ui/2605)), you can move it to `app/javascript/` and use newer JS features.

himdel replied:
Also, that `miqAjaxButton` is an anti-pattern now, with angular components, you should no longer rely on the server generating javascript to redirect you/show a flash message/whatever... instead, you should be using `render :json => ...` in ruby, and read that json in JS to do the right thing.

(But, it's not `miqAjaxButton(url, true)` so.. if the server-side logic to do that is not trivial, feel free to ignore this bit *for now*.)

tzumainn commented:
@martinpovolny Thanks for looking at this! I'm pretty new at JS so it's difficult for me unless I'm following existing patterns. Do you have an example of a controller being a component?

martinpovolny commented on app/controllers/application_controller/ci_processing.rb:
```diff
@@ -1049,6 +1051,8 @@ def process_vm_buttons(pfx)
     when "#{pfx}_associate_floating_ip"     then associate_floating_ip_vms
     when "#{pfx}_disassociate_floating_ip"  then disassociate_floating_ip_vms
+    when "#{pfx}_add_security_group"        then add_security_group_vms
+    when "#{pfx}_remove_security_group"     then remove_security_group_vms
```
Is this going to be relevant for multiple different entities?

What I mean: you are adding some "#{pfx}_action" here. It seems to be relevant for VMs, right? Is it going to be relevant for other entities?

Or is it just that you saw many of the VM button action being handled here?

If it's going to be relevant just for VMs then I'd prefen not to do the `pfx` magic.

But if you actually rewrite this to be pure client-side throught the API then you probably will not need to change this file (and the one below) at all.

tzumainn replied:
@martinpovolny hi! to be honest, I don't know what this code is for; would you suggest just removing this bit?

himdel commented on app/assets/javascripts/controllers/vm_cloud/vm_cloud_remove_security_group_form_controller.js:
```diff
@@ -0,0 +1,40 @@
+    vm.model = "vmCloudModel";
+    vm.saveable = miqService.saveable;
+    miqService.sparkleOn();
+    $http.get('/vm_cloud/remove_security_group_form_fields/' + vmCloudRemoveSecurityGroupFormId).then(function(response) {
```
Oh, also.. isn't this something that should be supported by the API?

I wonder if it even makes sense to add this functionality the old way - if it's already in the API, no need for this, and if it's not, why wasn't the API change done first?

tzumainn commented:
Depends on API PR: https://github.com/ManageIQ/manageiq-api/pull/81

tzumainn commented:
@himdel I'll be working on the angular changes next, but I've updated the code to use the API - can you let me know if you see anything I'm doing wrong in there? Thanks!

tzumainn commented:
@himdel thanks for the detailed notes on creating a js component! I think I did it right, let me know if I messed up somewhere.

himdel commented on app/assets/javascripts/components/vm_cloud/vm-cloud-add-security-group-component.js:
```diff
@@ -0,0 +1,61 @@
+  };
+
+  vm.resetClicked = function(angularForm) {
+    vm.CloudModel = angular.copy( vm.modelCopy );
```
this should be `vm.vmCloudModel`

himdel commented on app/assets/javascripts/components/vm_cloud/vm-cloud-add-security-group-component.js:
```diff
@@ -0,0 +1,61 @@
+        currentSecurityGroups = data.resources;      
+        API.get("/api/cloud_tenants/" + tenantId + "/security_groups?expand=resources&attributes=id,name").then(function(data) {
+          vm.security_groups = data.resources.filter(function (securityGroup) {
+            for (currentSecurityGroup of currentSecurityGroups) {
```
can't use `for..of` in `app/assets/javascripts/` .. you'd need to move this to `app/javascript/` or use `for..in` (or `.forEach`)

himdel commented on app/assets/javascripts/components/vm_cloud/vm-cloud-add-security-group-component.js:
```diff
@@ -0,0 +1,61 @@
+    vm.model = "vmCloudModel";
+    vm.saveable = miqService.saveable;
+    miqService.sparkleOn();
+    API.get("/api/vms/" + vm.recordId).then(function(data) {
```
Aah.. the nice thing about promises is that you don't have to nest them, you can always return a Promise instead...

This part should be:

```js
var tenantId;
var currentSecurityGroups;

API.get("/api/vms/" + vm.recordId)
  .then(function(data) {
    tenantId = data.cloud_tenant_id;

    return API.get("/api/vms/" + vm.recordId + "/security_groups?expand=resources&attributes=id,name");
  })
  .then(function(data) {
    currentSecurityGroups = data.resources;

    return API.get("/api/cloud_tenants/" + tenantId + "/security_groups?expand=resources&attributes=id,name");
  })
  .then(function(data) {
    vm.security_groups = data.resources.filter(function(securityGroup) {
      return ! _.find(currentSecurityGroups, { id: securityGroup.id });
    });

    vm.afterGet = true;
    vm.modelCopy = angular.copy(vm.vmCloudModel);
  })
  .catch(miqService.handleFailure);
```

himdel replied:
..and since the first 2 queries don't depend on each other, they can run in parallel..

```js
var currentSecurityGroups;

$q.all([
  API.get("/api/vms/" + vm.recordId),
  API.get("/api/vms/" + vm.recordId + "/security_groups?expand=resources&attributes=id,name"),
])
  .then(function(data) {
    var tenantId = data[0].cloud_tenant_id;
    currentSecurityGroups = data[1].resources;

    return API.get("/api/cloud_tenants/" + tenantId + "/security_groups?expand=resources&attributes=id,name");
  })
  .then(function(data) {
    vm.security_groups = data.resources.filter(function(securityGroup) {
      return ! _.find(currentSecurityGroups, { id: securityGroup.id });
    });

    vm.afterGet = true;
    vm.modelCopy = angular.copy(vm.vmCloudModel);
  })
  .catch(miqService.handleFailure);
```

himdel commented on app/assets/javascripts/components/vm_cloud/vm-cloud-add-security-group-component.js:
```diff
@@ -0,0 +1,61 @@
+        }).catch(miqService.handleFailure);
+      }).catch(miqService.handleFailure);
+    }).catch(miqService.handleFailure);
+    miqService.sparkleOff();
```
Ah, and it doesn't make sense to call `sparkleOff` directly from `$onInit`, since it will happen right away, not after the data actually arrives .. this should be inside that last `then` instead.

himdel commented on app/assets/javascripts/components/vm_cloud/vm-cloud-add-security-group-component.js:
```diff
@@ -0,0 +1,61 @@
+  vm.resetClicked = function(angularForm) {
+    vm.CloudModel = angular.copy( vm.modelCopy );
+    angularForm.$setPristine(true);
+    miqService.miqFlash("warn", "All changes have been reset");
```
`miqService.miqFlash("warn", __("All changes have been reset"))` .. i18n

himdel commented:
> Is there any need to have a Reset button? Doesn't seem to make sense to me. If you are adding something, you should be able to either Save or Cancel - unsure what Reset provides.

@tzumainn You'll probably just need to set `vm.newRecord = true;`

tzumainn commented:
@himdel thanks so much for the detailed suggestions! I've implemented them, let me know what you think!

@serenamarie125 I've removed the paging and the reset buttons, let me know if there's anything else!

tzumainn commented:
The API dependency is now merged.

serenamarie125 commented:
@tzumainn I'm sorry, i didn't see the update.  Can you add an additional mockup showing that the paging has been removed?  Thanks!

tzumainn commented:
@serenamarie125 sure, here it is!
![screenshot from 2017-10-18 22 27 18](https://user-images.githubusercontent.com/1280178/31751494-9e12c052-b453-11e7-9773-0a05b65af75f.png)

serenamarie125 commented:
@miq-bot add_labels ux/approved

serenamarie125 commented:
@miq-bot remove_labels ux/review

martinpovolny commented on app/controllers/mixins/actions/vm_actions/add_security_group.rb:
```diff
@@ -0,0 +1,100 @@
+            replace_right_cell
+          else
+            session[:flash_msgs] = @flash_array.dup
+            render :update do |page|
```
you can replace these 4 lines with a single `javascript_redirect` line and it will be much better that adding new RJS

martinpovolny commented on app/controllers/mixins/actions/vm_actions/add_security_group.rb:
```diff
@@ -0,0 +1,100 @@
+              @refresh_partial = "vm_common/add_security_group"
+            else
+              render :update do |page|
+                page << javascript_prologue
```
javascript_redirect

martinpovolny commented on app/controllers/mixins/actions/vm_actions/remove_security_group.rb:
```diff
@@ -0,0 +1,101 @@
+              @refresh_partial = "vm_common/remove_security_group"
+            else
+              render :update do |page|
+                page << javascript_prologue
```
javascript_redirect

martinpovolny commented on app/controllers/mixins/actions/vm_actions/remove_security_group.rb:
```diff
@@ -0,0 +1,101 @@
+          if @sb[:explorer]
+            replace_right_cell
+          else
+            session[:flash_msgs] = @flash_array.dup
```
javascript redirect

martinpovolny commented:
@tzumainn : Sorry to be the problem guy again.

I have put it some inline comments. Generally you should not neew any `render :update` when creating new stuff. 

Furthemore you should not be adding GET routes for descructive stuff or you may introduce a new security problem.

But the above has a limited significance in this PR. WHat I still don't understand why you are not using the API for the remove and add operations :-(

Can you explain that, please?

Why are you doing this?
```
+  vm.cancelClicked = function() {
 +    var url = '/vm_cloud/remove_security_group_vm/' + vm.recordId + '?button=cancel';
 +    miqService.miqAjaxButton(url);
 +  };
 +
 +  vm.addClicked = function() {
 +    var url = '/vm_cloud/remove_security_group_vm/' + vm.recordId + '?button=submit';
 +    miqService.miqAjaxButton(url, vm.vmCloudModel, { complete: false });
 +  };
 +}
```

Why do you need to go to the server and have Rails code for handling OK/cancel buttons?

Why don't you just do it all on the client and throught the API?

Is there something missing to do it right? 

How can we help you to really use the API for the new stuff?

I am sorry but I don't think we should merge this. We should really do this on the client so that we don't immediately introdce new code that will need rewriting.

We can setup a call if needed. Ping @AparnaKarve, @himdel : can you help @tzumainn make this pure client-side PR?

himdel commented:
@tzumainn For the angular-side, saving via the API can be seen in [ansible_credentials](https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/assets/javascripts/controllers/ansible_credentials/ansible_credentials_form_controller.js#L68-L78) or [tenant-component](https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/assets/javascripts/components/ops/tenant-component.js#L44-L70).

tzumainn commented:
depends on https://github.com/ManageIQ/manageiq-api/pull/137

tzumainn commented:
@martinpovolny I believe I've addressed your comments. Let me know if anything else should be changed.

martinpovolny commented:
@tzumainn : Thanks! It looks much better ;-) I really appreciate that you did these changes and worked with the API.

Moving forward we should figure how to avoid all the changes you did in:
```
app/controllers/mixins/actions/vm_actions/add_security_group.rb
app/controllers/mixins/actions/vm_actions/remove_security_group.rb
app/controllers/vm_common.rb
```
and some more. As this is really just routing from one Angular-enabled page to another one by doing a HTTP request. That should not be necessary and it should be simpler.

But we can do that in a separate PR.

@himdel, @h-kataria : what do you think?

@himdel, @ZitaNemeckova: Maybe we can take this for our Monday's "workshop" with @gildub and @aufi and make it into an example on how to progress with moving on the client. A blogpost maybe?

himdel commented on app/helpers/application_helper/toolbar/openstack_vm_cloud_center.rb:
```diff
@@ -69,6 +69,22 @@ class ApplicationHelper::Toolbar::OpenstackVmCloudCenter < ApplicationHelper::To
+          t = N_('Add a Security Group to this Instance'),
+          t,
+          :klass   => ApplicationHelper::Button::GenericFeatureButtonWithDisable,
+          :options => {:feature => :add_security_group}
```
This is probably not needed..  `instance_add_security_group` is the needed rbac feature, because of the toolbar button's identifier.

tzumainn replied:
Hi! I had thought that the preference in feature names was to use non-entity specific names when possible. It's certainly possible that we'll be adding security groups to other things in the future, so isn't add_security_group preferable?

martinpovolny replied:
Well until that happens I'd go with the simple case.

Other that that I think this is ready to merge.

tzumainn replied:
I understand the concern, but I've been explicitly told to do the opposite by reviewers who reviewed changes to the feature list in the supports_feature_mixin. If you look at https://github.com/ManageIQ/manageiq/blob/master/app/models/mixins/supports_feature_mixin.rb, none of the features are entity specific.

martinpovolny replied:
Oh well @himdel and I both mixed up RBAC feature and product feature.

martinpovolny commented:
I know I said "last" but before pressing the merge button I took a look at the CC and besided the similar code warning, that I don't push you to address there are 2 trivial ones:

 > Trailing spaces not allowed. NEW
 > Expected indentation of 8 spaces but found 7. NEW

Please, fix these 2.

tzumainn commented:
@miq-bot add_label pending/core

tzumainn commented:
@martinpovolny fixed the issues (and updated the API call in accordance to a request on the API PR). I'll let you know when that API PR is merged, thanks!

tzumainn commented:
@martinpovolny backend API PR is merged, so I think this should be ready! The test failures look unrelated... ?
